haskell-src/Concordium/Genesis/Data/P7.hs

```diff
@@ -0,0 +1,144 @@
+import Data.Word
+
+import Concordium.Common.Version
+import qualified Concordium.Crypto.SHA256 as Hash
+import qualified Concordium.Genesis.Data.Base as Base
+import qualified Concordium.Genesis.Data.BaseV1 as BaseV1
+import Concordium.Types
+
+-- | Initial genesis data for the P7 protocol version.
+data GenesisDataP7 = GDP7Initial
+    { -- | The immutable genesis parameters.
+      genesisCore :: !BaseV1.CoreGenesisParametersV1,
+      -- | Serialized initial block state.
+      --  NB: This block state contains some of the same values as 'genesisCore', and they should match.
+      genesisInitialState :: !(Base.GenesisState 'P7)
+    }
+    deriving (Eq, Show)
+
+-- | The regenesis represents a reset of the protocol with a new genesis block.
+--   This does not include the full new state, but only its hash.
+--
+--  The relationship between the new state and the state of the
+--  terminal block of the old chain should be defined by the
+--  chain update mechanism used.
+--
+--  There are two variants, one when migrating from 'P6' to 'P7' and
+--  one from 'P7' to 'P7'.
+data RegenesisP7
+    = GDP7Regenesis {genesisRegenesis :: !BaseV1.RegenesisDataV1}
+    | GDP7RegenesisFromP6
+        { genesisRegenesis :: !BaseV1.RegenesisDataV1,
+          genesisMigration :: ()
+        }
+    deriving (Eq, Show)
+
+-- | Deserialize genesis data in the V9 format.
+getGenesisDataV9 :: Get GenesisDataP7
+getGenesisDataV9 =
+    getWord8 >>= \case
+        0 -> do
+            genesisCore <- get
+            genesisInitialState <- get
+            return GDP7Initial{..}
+        _ -> fail "Unrecognized P7 genesis data type."
+
+getRegenesisData :: Get RegenesisP7
+getRegenesisData =
+    getWord8 >>= \case
+        1 -> do
+            genesisRegenesis <- get
+            return GDP7Regenesis{..}
+        2 -> do
+            genesisRegenesis <- get
+            genesisMigration <- get
+            return GDP7RegenesisFromP6{..}
+        _ -> fail "Unrecognized P7 regenesis data type."
+
+-- | Serialize genesis data in the V9 format.
+putGenesisDataV9 :: Putter GenesisDataP7
+putGenesisDataV9 GDP7Initial{..} = do
+    putWord8 0
+    put genesisCore
+    put genesisInitialState
+
+-- | Deserialize genesis data with a version tag. The expected version tag is 9
+--  and this must be distinct from version tags of other genesis data formats.
+getVersionedGenesisData :: Get GenesisDataP7
+getVersionedGenesisData =
+    getVersion >>= \case
+        9 -> getGenesisDataV9
+        n -> fail $ "Unsupported genesis data version for P7 genesis: " ++ show n
+
+-- | Serialize genesis data with a version tag.
+--  This will use the V9 format.
+putVersionedGenesisData :: Putter GenesisDataP7
+putVersionedGenesisData gd = do
+    putVersion 9
+    putGenesisDataV9 gd
+
+-- | Compute the block hash of the genesis block with the given genesis data.
+--  Every block hash is derived from a message that begins with the block slot,
+--  which is 0 for genesis blocks.
```

Comment:
We probably should have fixed this in P6 - it's too late now, but we can fix it in P7. We should make sure the hashing schemes for (re)genesis blocks are consistent with the general hashing scheme. In particular, we shouldn't be using slot numbers. This could wait until we revise the hashing scheme, but we should at least make a note to do it.

Reply:
I have added this issue for it https://github.com/Concordium/concordium-node/issues/1083